### ISSUE TITLE
Generates invalid output with certain properties

### ISSUE THREAD

**Issue by trueadm** [google/closure-compiler#3588]
The React team encountered an issue with Closure Compiler when compiling the React DOM bundle. It's quite specific, but it seems to somehow confuse the compiler into re-using a variable that is not safe to use given a given scenario.

https://github.com/facebook/react/pull/18670#issuecomment-623481140

Reproduction steps:

- Checkout branch from https://github.com/facebook/react/pull/18670#issuecomment-623481140
- Run `yarn build --type=NODE_PROD`
- Run `yarn test-prod-build ReactDOMServerIntegrationTextarea`. This will run the Jest test which shows the error.

This error should be seen:

`TypeError: Cannot read property 'focusedElem' of null`. It seems to be related to `JSCompiler_inline_result`. Which incorrectly gets re-used from a `null` value.

Specifically, the change that causes this is when we convert this logic:

```js
function prepareForCommit(containerInfo) {
  ...
}
...
prepareForCommit(root.containerInfo);
```

to:

```js
function prepareForCommit(containerInfo) {
  ...
  return null;
}
...
prepareForCommit(root.containerInfo) || null;
```

These are the flags we use for this particular build:

```js
const closureOptions = {
  compilation_level: 'SIMPLE',
  language_in: 'ECMASCRIPT_2015',
  language_out: 'ECMASCRIPT5_STRICT',
  env: 'CUSTOM',
  warning_level: 'QUIET',
  apply_input_source_maps: false,
  use_types_for_optimization: false,
  process_common_js_modules: false,
  rewrite_polyfills: false,
  inject_libraries: false,
  assume_function_wrapper: true,
};
```

**Comment by shicks** [comment]
It would be helpful if you could narrow down a smaller repro case.

**Comment by brad4d** [comment]
**tl;dr I don't think we can reasonably investigate this without a smaller repro that doesn't require building react.**

I tried to reproduce this today, but hit a roadblock.

```
$ yarn build --node=NODE_PROD
yarn run v1.22.4
$ node ./scripts/rollup/build.js --node=NODE_PROD
internal/modules/cjs/loader.js:583
    throw err;
    ^

Error: Cannot find module 'rollup'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (/usr/local/google/home/bradfordcsmith/b/155643260/react/scripts/rollup/build.js:3:16)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

# OK, I guess I have to do yarn install first...

$ yarn install
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
error eslint@6.8.0: The engine "node" is incompatible with this module. Expected version "^8.10.0 || ^10.13.0 || >=11.10.1". Got
 "10.9.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

I'm afraid that figuring out how to get react to build is beyond what we can reasonably do to investigate a bug. Sorry.

**Comment by trueadm** [comment]
@brad4d I'm sorry you weren't able to build React locally, let me help! If you install the latest stable Nodejs from here:

https://nodejs.org/en/download/

You should be able to run `yarn install` and repro the issue as per the original post.

Unfortunately, I haven't been able to make a smaller repro of this case, probably because of how large our bundles are, which might be what adds to the complexity of this particular problem.

**Comment by gaearon** [comment]
@trueadm I think your original description missed the `yarn` step before `yarn build`.

**Comment by brad4d** [comment]
@trueadm `yarn install` fails on my system. (See above).

I'm sorry, but I'm not going to go find a Windows machine in order to repro this issue.

**Comment by trueadm** [comment]
@brad4d This should work fine on a non Window machine. I personally use Mac/Ubuntu and the above works fine (unless you meant the inverse). The link above to install should also work for other OSs too, not just Windows. I can package up the build tomorrow for you, so there's no build step required, if that helps?

**Comment by gaearon** [comment]
Oh sorry I missed this part in your output.

```
$ yarn install
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
error eslint@6.8.0: The engine "node" is incompatible with this module. Expected version "^8.10.0 || ^10.13.0 || >=11.10.1". Got
 "10.9.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

If you run `yarn install --ignore-engines` it should work.

**Comment by gaearon** [comment]
Ping just in case — reusing variables when it's not safe seems like a scary bug. I totally understand if you'd rather not spend time on it due to a too large repro case though.

**Comment by blickly** [comment]
Oops, sorry this fell through a hole in our triage process for so long.

I think that the steps to run this do work after adding the `--ignore-engines` flag.  Namely, I'm able to run:

```bash
git clone https://github.com/trueadm/react.git
cd react/
yarn install --ignore-engines
yarn build --type=NODE_PROD
yarn test --prod --build ReactDOMServerIntegrationTextarea
```

That said, there's no `TypeError: Cannot read property 'focusedElem' of null` error. The final step passes, which I think is unexpected:
```
$ /usr/local/google/home/blickly/repos/node_modules/yarn/bin/yarn test --prod --build ReactDOMServerIntegrationTextarea
yarn run v1.22.10
$ node ./scripts/jest/jest-cli.js --prod --build ReactDOMServerIntegrationTextarea
$ NODE_ENV=production RELEASE_CHANNEL=experimental node ./scripts/jest/jest.js --config ./scripts/jest/config.build.js ReactDOMServerIntegrationTextarea

Running tests for default (experimental)...
 PASS  packages/react-dom/src/__tests__/ReactDOMServerIntegrationTextarea-test.js
  ReactDOMServerIntegrationTextarea
    ✓ renders a textarea with a value and an onChange with server string render (39ms)
    ✓ renders a textarea with a value and an onChange with server stream render (18ms)
    ✓ renders a textarea with a value and an onChange with clean client render (21ms)
    ✓ renders a textarea with a value and an onChange with client render on top of good server markup (35ms)
    ✓ renders a textarea with a value and an onChange with client render on top of bad server markup (23ms)
    ✓ renders a textarea with a value and readOnly with server string render (26ms)
    ✓ renders a textarea with a value and readOnly with server stream render (17ms)
    ✓ renders a textarea with a value and readOnly with clean client render (21ms)
    ✓ renders a textarea with a value and readOnly with client render on top of good server markup (37ms)
    ✓ renders a textarea with a value and readOnly with client render on top of bad server markup (24ms)
    ✓ renders a textarea with a value and no onChange/readOnly with server string render (16ms)
    ✓ renders a textarea with a value and no onChange/readOnly with server stream render (16ms)
    ✓ renders a textarea with a value and no onChange/readOnly with clean client render (21ms)
    ✓ renders a textarea with a value and no onChange/readOnly with client render on top of good server markup (42ms)
    ✓ renders a textarea with a value and no onChange/readOnly with client render on top of bad server markup (24ms)
    ✓ renders a textarea with a defaultValue with server string render (16ms)
    ✓ renders a textarea with a defaultValue with server stream render (17ms)
    ✓ renders a textarea with a defaultValue with clean client render (23ms)
    ✓ renders a textarea with a defaultValue with client render on top of good server markup (37ms)
    ✓ renders a textarea with a defaultValue with client render on top of bad server markup (23ms)
    ✓ renders a textarea value overriding defaultValue with server string render (16ms)
    ✓ renders a textarea value overriding defaultValue with server stream render (17ms)
    ✓ renders a textarea value overriding defaultValue with clean client render (26ms)
    ✓ renders a textarea value overriding defaultValue with client render on top of good server markup (37ms)
    ✓ renders a textarea value overriding defaultValue with client render on top of bad server markup (22ms)
    ✓ renders a textarea value overriding defaultValue no matter the prop order with server string render (16ms)
    ✓ renders a textarea value overriding defaultValue no matter the prop order with server stream render (16ms)
    ✓ renders a textarea value overriding defaultValue no matter the prop order with clean client render (20ms)
    ✓ renders a textarea value overriding defaultValue no matter the prop order with client render on top of good server markup (34ms)
    ✓ renders a textarea value overriding defaultValue no matter the prop order with client render on top of bad server markup (21ms)

Test Suites: 1 passed, 1 total
Tests:       30 passed, 30 total
Snapshots:   0 total
Time:        2.867s
Ran all test suites matching /ReactDOMServerIntegrationTextarea/i.
```

Does that mean that this issue has been fixed in the meantime?

**Comment by gaearon** [comment]
Did you check out this branch? https://github.com/facebook/react/pull/18670

I don’t believe the issue existed on master.

**Comment by blickly** [comment]
Ok, great, thanks for catching that!

Then the actual repro steps are:
```bash
git clone https://github.com/trueadm/react.git
cd react/
git checkout break-gcc
yarn install --ignore-engines
yarn build --type=NODE_PROD
yarn test-prod-build ReactDOMServerIntegrationTextarea
```

And that does give the expected error, namely (several) failed tests with:
```
 TypeError: Cannot read property 'focusedElem' of null
``

**Comment by gaearon** [comment]
Yep! Let me know if you need any more info. I don't have the time to reduce it right now but I can answer questions about the build/test setup etc. The build being used should be in `build/node_modules/react-dom/cjs/react-dom.production.min.js`.

**Comment by gaearon** [comment]
I think if you pass `--pretty` to the build command, it should disable the renaming pass and Prettify the output, which should make it easier to read and analyze.

**Comment by developit** [comment]
I had a go at trying to extract the issue into a standalone reproduction, but wasn't able to trigger the inlining bug:
https://gist.github.com/developit/0c919f6a2390dba1fa2e69d075738ae2

Not sure if it's useful, but here's roughly the difference between the working version and the broken one:
https://gist.github.com/developit/40189d836529562fd48954efc25e6c4f/revisions#diff-0052e4cc6e4061a1ecbc63f9cd9e149a2785dceda16e9e8eddc53f37de501ef7L16059

You can see that the `prepareForCommit()` and `getSelectionInformation()` functions (and a few others) are dropped entirely. That makes me wonder if this caused by Closure not knowing that certain DOM methods are side-effecting.

**Comment by brad4d** [comment]
I've just tried to follow the repro instructions posted by @blickly 
Sadly the world has moved on and the `break-gcc` branch doesn't work with the latest version of node.
I get a failure when I try to do the `yarn install`.

I tried just rebasing `break-gcc` onto `master`, but there are conflicts I don't know how to resolve.

**Comment by gaearon** [comment]
@brad4d

You got this error because `package.json` explicitly specifies "supported" Node versions as a list. Since Node 15 did not exist back then, it was not in the list. I've looked at the error stack to learn this, and then made this change to `package.json` to get `yarn` running again (https://github.com/trueadm/react/commit/911994149562c2699be04b99586fd61a5cddf3d8):

```diff
  "devEngines": {
-    "node": "8.x || 9.x || 10.x || 11.x || 12.x || 13.x"
+    "node": "*"
```

I've just pushed this fix for your convenience. I've verified that the incorrect output still reproduces with latest Node.

**Comment by gaearon** [comment]
If this is still cumbersome to follow I can take a stab at further extracting it. But I'm hoping that was the last missing step for you to get the repro running. Let me know if not!

**Comment by gaearon** [comment]
Here's another bug that might (or might not) be related: https://github.com/google/closure-compiler/issues/3788

**Comment by gaearon** [comment]
Is there still anything that prevents looking at this, or is the issue in the time constraints? I worry that if each repro attempt happens months after the report, there'll be something else that breaks, and we'll keep going in cycles. The repro is ready and should work today.

**Comment by brad4d** [comment]
OK, I've just reproduced this today.

All of the `"TypeError: Cannot read property 'focusedElem' of null"` error messages point one of 2 places.

```
    TypeError: Cannot read property 'focusedElem' of null

      51 |       if (forceHydrate) {
      52 |         ReactTestUtils.act(() => {
    > 53 |           ReactDOM.hydrate(reactElement, domElement);
         |                    ^
      54 |         });
      55 |       } else {
      56 |         ReactTestUtils.act(() => {

      at bk (build/node_modules/react-dom/cjs/react-dom.production.min.js:270:227)
```

OR

```
    TypeError: Cannot read property 'focusedElem' of null

      55 |       } else {
      56 |         ReactTestUtils.act(() => {
    > 57 |           ReactDOM.render(reactElement, domElement);
         |                    ^
      58 |         });
      59 |       }
      60 |       // We can't use the callback for resolution because that will not catch

      at bk (build/node_modules/react-dom/cjs/react-dom.production.min.js:270:227)
```

**Comment by brad4d** [comment]
@gaearon I want to modify the options that are passed to closure-compiler for the build, but it's not obvious to me where I would do that in the React build setup.

Also, what command can I give that would just mean "build the JS output" without running the tests?

**Comment by gaearon** [comment]
>All of the "TypeError: Cannot read property 'focusedElem' of null" error messages point one of 2 places.

It seems like the displayed location (`ReactDOM.render`/`ReactDOM.hydrate` call) is not helpful because it excludes the actual frame where the error happens inside the bundle. I presume maybe the test runner filters out the frames inside `node_modules`. The actual issue is in `build/node_modules/react-dom/cjs/react-dom.production.min.js`.

>I want to modify the options that are passed to closure-compiler for the build, but it's not obvious to me where I would do that in the React build setup.

Look for `closureOptions` in `scripts/rollup/build.js`. Note that it's being `Object.assign`ed with a few other options at the usage site, so look for its usage too.

>Also, what command can I give that would just mean "build the JS output" without running the tests?

```
# This only does the build:
yarn build --type=NODE_PROD

# This only runs the tests:
yarn test-prod-build ReactDOMServerIntegrationTextarea
```

----

I pushed one more commit. With that commit, you can get a faster build for easier iteration.

**After you `git pull`,** you should be able to replace `yarn build --type=NODE_PROD` (which takes a lot of time) with this:

```
yarn build react/index,react-dom/index,react-dom/test-utils,react-dom/server,scheduler --type=NODE_PROD
```

(builds much faster)

You'd still need to run `yarn test-prod-build ReactDOMServerIntegrationTextarea` after each build to actually run the tests.

**Additionally**, you can use the `--pretty` flag when building to disable the minification pass. This makes it a lot more readable. So in that case the command would be:

```
yarn build react/index,react-dom/index,react-dom/test-utils,react-dom/server,scheduler --type=NODE_PROD --pretty
```

I verified that running `yarn test-prod-build ReactDOMServerIntegrationTextarea` after that still produces the error, but the bundle in `build/node_modules/react-dom/cjs/react-dom.production.min.js` is now human readable.

----

To sum up, my suggested workflow is:

```
git pull
```

And then

```
# To build:
yarn build react/index,react-dom/index,react-dom/test-utils,react-dom/server,scheduler --type=NODE_PROD --pretty

# To test:
yarn test-prod-build ReactDOMServerIntegrationTextarea
```

And manually inspect `build/node_modules/react-dom/cjs/react-dom.production.min.js` to see the code.

**Comment by brad4d** [comment]
Thx for the detailed instructions.

I'm seeing stuff that makes no sense to me at all.

With your new build instructions when I look in `react-dom.production.min.js` there is only 1 location where a `focusedElem` property is accessed.

```javascript
    JSCompiler_inline_result = current.focusedElem;
```

That should narrow it down, right?
With some careful grepping I believe the reference to `focusedElem` we're interested in is in `ReactInputSelections.js` on line 119.

What's more I can see in  `react-dom.production.min.js` that 
1. `current` gets assigned the value from `selectionInformation`
2. `selectionInformation` is only ever assigned the value `null`.

Hmmm.

I'm also really puzzled about why I see `JSCompiler_inline_result` used all over the place.
That is only ever supposed to be a prefix to which additional text is added for uniqueness.

The string is defined here.
https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/ExpressionDecomposer.java#L787

And it is only used here, where suffixes are added.
https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/ExpressionDecomposer.java#L808

I cannot explain that. Could you be using some old version of the compiler?
(Though I don't think "old compiler" really would explain it. We have **always** generated unique suffixes AFAIK.)

**Comment by brad4d** [comment]
The `restoreSelection` method containing the problem reference to `focusedElem` is only called in one `ReactDOMHostConfig.js` on line 270, where it is passed the variable `selectionInformation`.

The only assignment to `selectionInformation` that isn't `null` is on line 255 where it is assigned `getSelectionInformation()` inside of the `prepareForCommit()` function.

AFAICT the version of `prepareForCommit()` in `ReactDOMHostConfig.js` is never called anywhere.
The only calls to it I can find are importing it from `ReactFiberHostConfig.js`

Is it possible that there are actually, intentionally, **no** calls to `prepareForCommit()` in the set of files you're passing to closure-compiler, because you expect it to be called from something outside of the compilation unit? If so, you need to define externs to prevent it from being removed.

You may find some useful advice on that here.
https://github.com/google/closure-compiler/wiki/How-do-I-make-closure-compiler-stop-breaking-my-API-by-renaming-things%3F

**Comment by gaearon** [comment]
>AFAICT the version of prepareForCommit() in ReactDOMHostConfig.js is never called anywhere. The only calls to it I can find are importing it from ReactFiberHostConfig.js

Yes, but `ReactFiberHostConfig` is being replaced by `ReactDOMHostConfig` at the build time before it is passed to the Google Closure Compiler. So any import of `ReactFiberHostConfig` actually becomes an import from `ReactDOMHostConfig`.

>It seems the problem is related to a need for the project to use externs to prevent the removal of symbols that are part of its published API.

No, this is incorrect. There are no external calls to it. I would not have filed this issue if there were external calls.

**Comment by gaearon** [comment]
>I'm also really puzzled about why I see JSCompiler_inline_result used all over the place. That is only ever supposed to be a prefix to which additional text is added for uniqueness.

Yes, I also find this behavior of Google Closure Compiler confusing. This is why I filed this issue in the Google Closure Compiler repository.

**Comment by gaearon** [comment]
Looks like I'm wrong on this one. I've extracted the standalone case, and it appears that the input passed to GCC is *already* wrong. I suspect that it's the earlier Rollup pass that causes the issue. Since both Rollup and GCC do optimizations.

So you were right to close the issue. My apologies.

**Comment by gaearon** [comment]
For posterity, here's the likely cause: https://github.com/rollup/rollup/issues/4455
